testing/test_hookcaller.py

```diff
@@ -213,3 +228,28 @@ def hello(self, arg):
     assert not hasattr(hook, "world")
     pm.unregister(plugin)
     assert hook.hello(arg=3) == []
+
+
+def test_hookrelay_registration_by_specname(pm):
+    """Verify hook caller instances may also be registered by specifying a
+    specname option to the hookimpl"""
+
+    class Api(object):
+        @hookspec
+        def hello(self, arg):
+            "api hook 1"
+
+    pm.add_hookspecs(Api)
+    hook = pm.hook
+    assert hasattr(hook, "hello")
+    assert len(pm.hook.hello.get_hookimpls()) == 0
+
+    class Plugin(object):
+        @hookimpl(specname="hello")
+        def foo(self, arg):
+            return arg + 1
+
+    plugin = Plugin()
+    pm.register(plugin)
+    out = hook.hello(arg=3)
```

Comment:
This is good for the expected "matching" function signature case but what about the case where it mismatches?

Should we maybe also check that a registration of an impl with an incorrect signature fails at `pm.register()`?

Reply:
added it.  Also added an assertion that `check_pending` fails if `specname` is provided by doesn't have a corresponding spec (even if the function name itself matches a spec).